cpp/tests/io/parquet_test.cpp

```diff
@@ -223,25 +223,73 @@ bool read_footer(std::unique_ptr<cudf_io::datasource>& source,
 
   // returns true on success
   return cp.read(file_meta_data);
 }
 
+// read column index from datasource at location indicated by chunk,
+// parse and return as a ColumnIndex struct.
+// throws cudf::logic_error if the chunk data is invalid.
+cudf::io::parquet::ColumnIndex read_column_index(std::unique_ptr<cudf::io::datasource>& source,
+                                                 const cudf::io::parquet::ColumnChunk& chunk)
+{
+  CUDF_EXPECTS(chunk.column_index_offset > 0, "Cannot find column index");
+  CUDF_EXPECTS(chunk.column_index_length > 0, "Invalid column index length");
+
+  cudf::io::parquet::ColumnIndex colidx;
+  const auto ci_buf = source->host_read(chunk.column_index_offset, chunk.column_index_length);
+  cudf::io::parquet::CompactProtocolReader cp(ci_buf->data(), ci_buf->size());
+  CUDF_EXPECTS(cp.read(&colidx), "Cannot parse column index");
+  return colidx;
+}
+
+// read offset index from datasource at location indicated by chunk,
+// parse and return as an OffsetIndex struct.
+// throws cudf::logic_error if the chunk data is invalid.
+cudf::io::parquet::OffsetIndex read_offset_index(std::unique_ptr<cudf::io::datasource>& source,
+                                                 const cudf::io::parquet::ColumnChunk& chunk)
+{
+  CUDF_EXPECTS(chunk.offset_index_offset > 0, "Cannot find offset index");
+  CUDF_EXPECTS(chunk.offset_index_length > 0, "Invalid offset index length");
+
+  cudf::io::parquet::OffsetIndex offidx;
+  const auto oi_buf = source->host_read(chunk.offset_index_offset, chunk.offset_index_length);
+  cudf::io::parquet::CompactProtocolReader cp(oi_buf->data(), oi_buf->size());
+  CUDF_EXPECTS(cp.read(&offidx), "Cannot parse offset index");
+  return offidx;
+}
+
 // parse the statistics_blob on chunk and return as a Statistics struct.
 // throws cudf::logic_error if the chunk statistics_blob is invalid.
-cudf_io::parquet::Statistics parse_statistics(const cudf_io::parquet::ColumnChunk& chunk)
+cudf::io::parquet::Statistics parse_statistics(const cudf::io::parquet::ColumnChunk& chunk)
 {
   auto& stats_blob = chunk.meta_data.statistics_blob;
   CUDF_EXPECTS(stats_blob.size() > 0, "Invalid statistics length");
 
-  cudf_io::parquet::Statistics stats;
-  cudf_io::parquet::CompactProtocolReader cp(stats_blob.data(), stats_blob.size());
+  cudf::io::parquet::Statistics stats;
+  cudf::io::parquet::CompactProtocolReader cp(stats_blob.data(), stats_blob.size());
   CUDF_EXPECTS(cp.read(&stats), "Cannot parse column statistics");
   return stats;
 }
 
+// read page header from datasource at location indicated by page_loc,
+// parse and return as a PageHeader struct.
+// throws cudf::logic_error if the page_loc data is invalid.
+cudf::io::parquet::PageHeader read_page_header(std::unique_ptr<cudf::io::datasource>& source,
+                                               const cudf::io::parquet::PageLocation& page_loc)
+{
+  CUDF_EXPECTS(page_loc.offset > 0, "Cannot find page header");
```

Comment:
The page header can't be the first byte in the datasource?

Reply:
Nope.  There's at least 4 bytes of header before the first data page.  Guess it could be >= 4